File: cpp/test/test_parser.cpp
// This file is part of an MIT-licensed project: see LICENSE file or README.md for details.
// Copyright (c) 2025 Ian Thomas

#include "expression_parser/parser.h"
#include "catch_amalgamated.hpp"
#include "test_utils.h"
#include <fstream>

using namespace ExpressionParser;

TEST_CASE( "Simple") {

    Parser parser;

    auto expression = parser.Parse("get_name()=='fred' and counter>0 and 5/5.0!=0");

    Context context;
    context["get_name"] = make_function_wrapper([]() -> std::string {
        return "fred";
    });
    context["counter"] = 1;

    std::any result = expression->Evaluate(context);

    REQUIRE(std::any_cast<bool>(result) == true);
}

TEST_CASE( "MatchOutput") {
    std::string source = loadTestFile("Parse.txt");

    std::vector<std::string> lines;
    std::istringstream iss(source);
    std::string line;
    while (std::getline(iss, line)) {
        lines.push_back(line);
    }

    Context context;
    context["C"] = 15;
    context["D"] = false;
    context["get_name"] = make_function_wrapper([]() -> std::string {
        return "fred";
    });
    context["end_func"] = make_function_wrapper([]() -> bool {
        return true;
    });
    context["whisky"] = make_function_wrapper([](const std::string &id, double n) -> std::string {
        int inum = static_cast<int>(n);
        return std::to_string(inum) + "whisky_" + id;
    });
    context["counter"] = 1;

    Parser parser;
    std::vector<std::string> processedLines;
    for (const auto &l : lines) {
        if (l.rfind("//", 0) == 0) {
            processedLines.push_back(l);
            continue;
        }

        // Append the original line wrapped in quotes.
        processedLines.push_back("\"" + l + "\"");
        try {
            auto node = parser.Parse(l);
            processedLines.push_back(node->DumpStructure());

            std::vector<std::string> dumpEval;
            node->Evaluate(context, &dumpEval);

            std::string evalOutput = joinStrings(dumpEval, "\n");
            processedLines.push_back(evalOutput);
        }
        catch (const std::exception &ex) {
            processedLines.push_back(ex.what());
        }
        processedLines.push_back("");
    }

    std::string output = joinStrings(processedLines, "\n");

    /*std::ofstream outFile("output.txt");
    if (!outFile) {
        throw std::runtime_error("Failed to open output.txt for writing.");
    }
    outFile << output;
    outFile.close();*/

    std::string match = loadTestFile("Parse-Output.txt");
    REQUIRE(match == output);
}

File: python/tests/test_parser.py
# This file is part of an MIT-licensed project: see LICENSE file or README.md for details.
# Copyright (c) 2025 Ian Thomas

import unittest
import sys
import os

sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "../src")))

from expression_parser.parser import Parser

class TestParser(unittest.TestCase):

    def setUp(self):
        self.maxDiff = None  # Allow full diff output for every test case

    def _load_file(self, file_name):
        try:
            with open(f"../tests/{file_name}", "r", encoding="utf-8") as file:
                return file.read()
        except Exception as e:
            self.fail(f"Error loading {file_name}: {e}")

    def test_simple(self):
        
        parser = Parser()
        expression = parser.parse("get_name()=='fred' and counter>0 and 5/5.0!=0")

        context = {
            "get_name":lambda: "fred",
            "counter": 1
        }

        result = expression.evaluate(context)

        self.assertEqual(result, True, "Expression should return True")


    def test_parse(self):

        source = self._load_file("Parse.txt")

        # Split into lines
        lines = source.splitlines()

        context = {
            "C":15,
            "D":False,
            "get_name":lambda: "fred",
            "end_func":lambda: True,
            "whisky": lambda id, n: str(int(n))+"whisky_"+id,
            "counter":1
        }

        parser = Parser()

        processed_lines = []
        for line in lines:
            if (line.startswith("//")):
                processed_lines.append(line)
                continue

            processed_lines.append(f'"{line}"')
            try:
                node = parser.parse(line)
                processed_lines.append(node.dump_structure())
                dump_eval = []
                node.evaluate(context, dump_eval)
                processed_lines.append("\n".join(dump_eval))
                
            except Exception as e:
                processed_lines.append(str(e))

            processed_lines.append("")

        output = "\n".join(processed_lines)

        #print(output)

        match = self._load_file("Parse-Output.txt")
        self.assertMultiLineEqual(match, output)

if __name__ == "__main__":
    unittest.main()